lib/resolve-idx-refs.js

```diff
@@ -0,0 +1,36 @@
+'use strict'
+
+const scanner = require('object-scan')
+const toPath = require('lodash/toPath')
+const get = require('lodash/get')
+
+// For all items in `tree` matching a path selector specified
+// in `selector`, call `onRef` with the item, its parent and
+// the path to the item.
+// Example:
+// tree: {foo: [{bar: 1}], hey: {there: [{bar: 2}]}}
+// selector: **[*].bar
+const findIdxRefs = (tree, selector, onRef) => {
+	const scan = scanner([selector])
+	for (const pathStr of scan(tree)) {
+		const path = toPath(pathStr)
+		const val = get(tree, path)
+		const parent = get(tree, path.slice(0, -1), {})
+		onRef(val, parent, path)
+	}
+}
+
+// For all items in `tree` matching a path selector specified
+// in `selector`:
+// - interpret the item as an index in `source`
+// - set the parent's field `prop` to `source[item]`
+const resolveIdxRefs = (tree, selector, source, prop) => {
+	findIdxRefs(tree, selector, (idx, parent) => {
+		if ('number' === typeof idx) parent[prop] = source[idx]
```

Comment:
Ok. If I understood everything correctly, this code does what it should, however I'm - to put it mildly - not the biggest fan of mutating the original object. I think it would be cleaner (and easier to test) to have a function that extracts a property from the original tree without modifying it, but that's just a style question, and I don't want to block this PR for this.

Reply:
> I think it would be cleaner (and easier to test) to have a function that extracts a property from the original tree without modifying it [...].

I agree in theory. In practice, letting `parse/common.js` use `findIdxRefs` directly would just move boilerplate code. Do you agree?

Reply:
I agree, but this would make it much easier to reason about what happens. In `parse/common`, I kind of have the *bigger picture* and can tell what is happening, while in this function here, I don't really see the bigger picture, but I indirecty see the function changing something in it.

Reply:
Changed it in a3f0cee.